File: langs/python/src/dotlyte/cli/check.py
"""``dotlyte check`` — Validate .env files against schema."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

from dotlyte.errors import DotlyteError


def run_check(
    *,
    schema_path: str | None = None,
    env_file: str | None = None,
    strict: bool = False,
) -> None:
    """Validate .env / config files, optionally against a JSON schema.

    Args:
        schema_path: Path to a JSON schema file. Auto-detected if ``None``.
        env_file: Explicit .env file to validate.
        strict: Reject keys not present in the schema.

    """
    print("🔍 dotlyte check\n")

    if schema_path is None:
        schema_path = _find_schema_file()

    if schema_path is None:
        print("⚠️  No schema file found. Performing basic validation...\n")
        _basic_check(env_file)
        return

    print(f"Schema: {schema_path}")

    # Load schema
    try:
        content = Path(schema_path).read_text(encoding="utf-8")
        schema_data: dict[str, Any] = json.loads(content)
    except (OSError, json.JSONDecodeError) as exc:
        raise DotlyteError(f"Failed to read schema file: {schema_path}: {exc}") from exc

    from dotlyte.loader import load
    from dotlyte.validator import validate_schema

    config = load(files=[env_file] if env_file else None)
    data = config.to_dict()
    violations = validate_schema(data, schema_data, strict)

    if not violations:
        print("\n✅ All checks passed!\n")
        return

    print(f"\n❌ {len(violations)} issue(s) found:\n")
    for v in violations:
        print(f"  ✗ {v.key}: {v.message}")
    print()
    sys.exit(1)


def _basic_check(env_file: str | None = None) -> None:
    """Perform basic validation when no schema is available."""
    files = [".env", ".env.local", ".env.example"]
    if env_file:
        files.insert(0, env_file)

    found = False
    for file in files:
        path = Path(file).resolve()
        if path.is_file():
            found = True
            print(f"  ✓ {file} exists")

            content = path.read_text(encoding="utf-8")
            keys: dict[str, int] = {}
            for line in content.splitlines():
                trimmed = line.strip()
                if not trimmed or trimmed.startswith("#"):
                    continue
                eq_idx = trimmed.find("=")
                if eq_idx > 0:
                    key = trimmed[:eq_idx].strip()
                    keys[key] = keys.get(key, 0) + 1

            dupes = [k for k, count in keys.items() if count > 1]
            if dupes:
                print(f"    ⚠️  Duplicate keys: {', '.join(dupes)}")

    if not found:
        print("  ⚠️  No .env files found in current directory")

    print("\n✅ Basic check complete\n")


def _find_schema_file() -> str | None:
    """Search for common schema file names."""
    candidates = [
        "env-schema.json",
        ".env.schema.json",
        "dotlyte.schema.json",
    ]
    for c in candidates:
        if Path(c).resolve().is_file():
            return c
    return None

File: langs/javascript/src/cli/commands/check.ts
/**
 * `dotlyte check` — Validate .env files against schema.
 *
 * Usage:
 *   dotlyte check [--schema <file>] [--env <file>]
 *
 * Validates .env/config files and exits with code 1 on failure (CI-friendly).
 */

import { existsSync, readFileSync } from "node:fs";
import { resolve } from "node:path";
import { load } from "../../loader.js";
import { validateSchema } from "../../validator.js";
import type { DotlyteSchema } from "../../validator.js";

export async function check(args: string[]): Promise<void> {
  const schemaPath = getArg(args, "--schema") ?? findSchemaFile();
  const envFile = getArg(args, "--env");
  const strict = args.includes("--strict");

  console.log("🔍 dotlyte check\n");

  if (!schemaPath) {
    console.log("⚠️  No schema file found. Performing basic validation...\n");
    basicCheck(envFile);
    return;
  }

  console.log(`Schema: ${schemaPath}`);

  // Load schema
  let schema: DotlyteSchema;
  try {
    const content = readFileSync(resolve(schemaPath), "utf-8");
    schema = JSON.parse(content) as DotlyteSchema;
  } catch {
    throw new Error(`Failed to read schema file: ${schemaPath}`);
  }

  // Load config
  const config = load({
    files: envFile ? [envFile] : undefined,
    schema: undefined, // We'll validate manually for better reporting
  });

  const data = config.toObject();
  const violations = validateSchema(data, schema, strict);

  if (violations.length === 0) {
    console.log("\n✅ All checks passed!\n");
    return;
  }

  console.log(`\n❌ ${violations.length} issue(s) found:\n`);
  for (const v of violations) {
    console.log(`  ✗ ${v.key}: ${v.message}`);
  }
  console.log();
  process.exit(1);
}

function basicCheck(envFile?: string): void {
  const files = [".env", ".env.local", ".env.example"];
  if (envFile) files.unshift(envFile);

  let found = false;
  for (const file of files) {
    const path = resolve(file);
    if (existsSync(path)) {
      found = true;
      console.log(`  ✓ ${file} exists`);

      // Check for duplicate keys
      const content = readFileSync(path, "utf-8");
      const keys = new Map<string, number>();
      for (const line of content.split("\n")) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith("#")) continue;
        const eqIdx = trimmed.indexOf("=");
        if (eqIdx > 0) {
          const key = trimmed.slice(0, eqIdx).trim();
          keys.set(key, (keys.get(key) ?? 0) + 1);
        }
      }

      const dupes = [...keys.entries()].filter(([, count]) => count > 1);
      if (dupes.length > 0) {
        console.log(`    ⚠️  Duplicate keys: ${dupes.map(([k]) => k).join(", ")}`);
      }
    }
  }

  if (!found) {
    console.log("  ⚠️  No .env files found in current directory");
  }

  console.log("\n✅ Basic check complete\n");
}

function findSchemaFile(): string | null {
  const candidates = [
    "env-schema.json",
    ".env.schema.json",
    "dotlyte.schema.json",
    "env-schema.ts",
  ];
  for (const c of candidates) {
    if (existsSync(resolve(c))) return c;
  }
  return null;
}

function getArg(args: string[], flag: string): string | undefined {
  const idx = args.indexOf(flag);
  return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
